doc/man/doveadm-indexer.1.in
.\" Copyright (c) 2022 Dovecot authors, see the included COPYING file
.TH DOVEADM\-INDEXER 1 "2022-02-02" "Dovecot v2.3" "Dovecot"
.SH NAME
doveadm\-indexer \- Commands related to managing the indexer process
.\"------------------------------------------------------------------------
.SH SYNOPSIS
.BR doveadm " [" \-Dv ]
[\fB\-f\fP \fIformatter\fP]
.BI indexer \ command
.RI [ OPTIONS ]\ [ ARGUMENTS ]
\"------------------------------------------------------------------------
.SH DESCRIPTION
.B doveadm indexer
can be used to manage the indexer process.
.\"------------------------------------------------------------------------
@INCLUDE:global-options-formatter@
.\"------------------------------------------------------------------------
.SH COMMANDS
.SS indexer add
.B doveadm indexer add
[\fB\-h\fP]
[\fB\-n\fP \fImax_recent\fP]
.I user
.I mailbox
.PP
Add indexing request for the given
.I user
and the
.I mailbox
to the indexer queue. 
It works the same as the
.B doveadm index -q
command.
.PP
.\"-------------------------------------
.TP
.B \-h
Add the indexing request to the head of the queue. By default the request is
added to the tail of the queue.
.\"-------------------------------------
.TP
.BI \-n \ max_recent
An integer value, which specifies the maximum number of \(rsRecent
messages in mailboxes.
If the mailbox contains more than
.I max_recent
messages with \(rsRecent flag set, the mailbox will not be indexed.
This may be useful to avoid unnecessary indexing for large mailboxes that
are never opened.
.\"------------------------------------------------------------------------
.SS indexer remove
.B doveadm indexer remove
.I user_mask
.PP
Remove all indexer requests for the matching
.IR user_mask .
It\(aqs possible to use wildcards. Requests that are currently processed by
indexer-worker are not listed; use
.B doveadm kick
instead to kick them.
.\"------------------------------------------------------------------------
.SS indexer list
.B doveadm indexer list
.IR user_mask " [" mailbox_mask ]
.PP
List all the queued indexing requests matching the
.I user_mask
(and the
.IR mailbox_mask ).
It\(aqs possible to use wildcards. Requests that are currently processed by
indexer-worker are not listed; use
.B doveadm who
instead to see them.
.\"------------------------------------------------------------------------
@INCLUDE:reporting-bugs@
.\"------------------------------------------------------------------------
.SH SEE ALSO
.BR doveadm (1)
.BR doveadm-index (1)


NEWS
v2.3.17.1 2021-12-07  Aki Tuomi <aki.tuomi@open-xchange.com>

	- dsync: Add back accidentically removed parameters.
	- lib-ssl-iostream: Fix assert-crash when OpenSSL returned syscall error
	  without errno.
	- master: Dovecot failed to start if ssl_ca was too large.

v2.3.17 2021-10-28  Aki Tuomi <aki.tuomi@open-xchange.com>

	* Dovecot now logs a warning if time seems to jump forward at least
	  100 milliseconds.
	* dict: Lines logged by the dict process now contain the dict name as
	  the prefix.
	* lib-index: mail_cache_fields, mail_always_cache_fields and
	  mail_never_cache_fields now verifies that the listed header names are
	  valid. Especially the UTF8 "–" character has sometimes been wrongly
	  used instead of the ASCII "-".
	+ *-login: Added login_proxy_rawlog_dir setting to capture
	  rawlogs between proxy and backend.
	+ dict: The server process now keeps the last 10 idle dict backends
	  cached for maximum of 30 seconds. Practically this acts as a
	  connection pool for dict-redis and dict-ldap. Note that this doesn't
	  affect dict-sql, because it already had its own internal cache.
	+ doveadm: New stats add/remove commands added to support changing the
	  metrics configuration on runtime.
	+ lazy_expunge: Added lazy_expunge_exclude settings to disable
	  lazy_expunge for specific folders. \Special-use flags can be used as
	  folder names.
	+ lib-lua: Added a new helper function dovecot.restrict_global_variables()
	  to disable or enable defining new global variables.
	- LAYOUT=index List index rebuild was missing.
	- LAYOUT=index: Duplicate GUIDs were not detected.
	- acl: When using acl_ignore_namespace Dovecot attempted to access or
	  create dovecot-acl-list even when the namespace should have been
	  ignored. For virtual namespaces this could have yielded errors about
	  "Read-only file system" or "Permission denied".
	- auth: Setting the "master" passdb field to empty value would
	  cause proxying to fail with an authentication error.
	  Now an empty "master" field is ignored.
	- doveadm-server: Duplicate error lines were sent for failed commands.
	  This didn't normally cause visible problems, except when using
	  wildcards in usernames or -A parameter to go through multiple users.
	- doveadm-server: Logs written by doveadm-server were often missing log
	  prefixes, especially mail_log_prefix for mail commands. Logs sent to
	  doveadm TCP client were also missing log prefixes.
	- doveadm: v2.3 regression: batch command always crashes.
	- doveadm: v2.3.11 regression: Commands failed if ssl_cert or
	  ssl_key files weren't readable by the user running doveadm, even
	  though doveadm didn't actually use these settings
	- imap-hibernate: Process may crash at deinit:
	  Panic: file ioloop.c: line 928 (io_loop_destroy): assertion failed:
	  (ioloop->cur_ctx == NULL).
	- imap: Using imap_fetch_failure=no-after can cause assert-crash
	  with some IMAP commands if reading the mail fails (e.g. wrong cached
	  mail size). Fixes:
	  Panic: file index-mail-headers.c: line 198 (index_mail_parse_header_init):
	  assertion failed: (!mail->data.header_parser_initialized)
	- imap: v2.3.10 regression: When using INDEXPVT to enable private
	  \Seen flags (for shared or public namespaces) the STORE command did
	  not send untagged replies for the \Seen flag changes.
	- imap: v2.3.15 regression: If PREVIEW/SNIPPET is not the final FETCH
	  option in the command, the IMAP FETCH response is broken.
	- imap: v2.3.15 regression: MOVE command leaks mailbox if it can't be
	  opened and crashes at deinit:
	  Panic: file mail-user.c: line 229 (mail_user_deinit): assertion failed:
	  ((*user)->refcount == 1).
	- imapc: Copying nonexistent mail via imapc could have crashed. Fixes:
	  Panic: file mail-storage.c: line 2385 (mailbox_transaction_commit_get_changes):
	  assertion failed: (ret < 0 || seq_range_count(&changes_r->saved_uids) == save_count ||
	  array_count(&changes_r->saved_uids) == 0).
	- indexer: v2.3.15 regression: Process crashes if indexer-client
	  disconnects while it's waiting for command reply. This happened for
	  example if IMAP SEARCH triggered long fts indexing and the IMAP
	  client disconnected while waiting for the reply.
	- indexer: v2.3.15 regression: Process may have crashed in some situations.
	- indexer: v2.3.15 regression: indexer-worker processes may not have
	  reached the process_limit in some situations, possibly even using just
	  one indexer-worker process even though there were many indexing
	  requests queued.
	- lib-compression: Reading lz4 compressed mdbox mails may crash. Fixes:
	  Panic: file istream.c: line 345 (i_stream_read_memarea):
	  assertion failed: (!stream->blocking).
	- lib-compression: bench-compress crashes due to xz being read-only.
	- lib-lua: Fix linking libdict_lua for non-GNU linkers when Lua support
	  is disabled.
	- lib-mail: There was no limit on how large an email header name could be.
	  Processable header names are now limited to 1000 bytes.
	- lib-oauth2: Dovecot disallowed JWT tokens if their validity time was
	  older than token creation time (nbf < iat).
	- lib-storage: Reduce memory footprint of certain storage operations.
	- lib-storage: When listing mailboxes with storage name escape
	  characters (^ or .) as part of the mailbox name, the listing could
	  show corrupted mailbox names. Due to an issue in handling escaped
	  parent folders, the listing of other mailbox names would become
	  corrupted by prepending parts of the previously listed mailboxes
	  parent folder as prefix to the actual mailbox names. The corruption
	  can occur when using LAYOUT=INDEX and maildir or obox, or when using
	  the listescape plugin.
	- mail-crypt: Fix "-O" argument for "doveadm mailbox cryptokey password"
	  command to be a boolean, and not expect a string.
	- submission-login: Add support for not authenticating to next hop in
	  submission proxying.
	- submission-login: EHLO was not sent again after XCLIENT when doing

src/doveadm/doveadm-indexer.c
/* Copyright (c) 2022 Dovecot authors, see the included COPYING file */

#include "lib.h"
#include "str.h"
#include "strescape.h"
#include "connection.h"
#include "istream.h"
#include "ostream.h"
#include "doveadm.h"
#include "doveadm-print.h"

#include <unistd.h>

#define INDEXER_SOCKET_NAME "indexer"

static struct istream *indexer_send_cmd(const char *cmd)
{
	const char *path = t_strconcat(doveadm_settings->base_dir,
				       "/"INDEXER_SOCKET_NAME, NULL);
	const struct connection_settings set = {
		.service_name_out = "indexer-client",
		.service_name_in = "indexer-server",
		.major_version = 1,
		.minor_version = 0,
	};
	const char *error;
	struct istream *input;
	struct ostream *output;
	if (doveadm_blocking_connect(path, &set, &input, &output, &error) < 0)
		i_fatal("%s", error);
	o_stream_nsend_str(output, t_strconcat(cmd, "\r\n", NULL));
	if (o_stream_flush(output) < 0)
		i_fatal("write(indexer) failed: %s", o_stream_get_error(output));
	o_stream_unref(&output);
	return input;
}

static struct istream *
indexer_send_cmd_with_args(const char *cmd, const char *const *args)
{
	string_t *str = t_str_new(128);

	str_append(str, cmd);
	if (args != NULL) {
		for (unsigned int i = 0; args[i] != NULL; i++) {
			str_append_c(str, '\t');
			str_append_tabescaped(str, args[i]);
		}
	}
	return indexer_send_cmd(str_c(str));
}

static void cmd_indexer_add(struct doveadm_cmd_context *cctx)
{
	const char *user, *mailbox, *line;
	int64_t max_recent;
	bool head;

	if (!doveadm_cmd_param_bool(cctx, "head", &head))
		head = FALSE;
	if (!doveadm_cmd_param_int64(cctx, "max-recent", &max_recent))
		max_recent = 0;
	if (!doveadm_cmd_param_str(cctx, "user", &user) ||
	    !doveadm_cmd_param_str(cctx, "mailbox", &mailbox))
		help_ver2(&doveadm_cmd_indexer_add);

	const char *cmd = head ? "PREPEND" : "APPEND";
	const char *const args[] = {
		"0", user, mailbox, dec2str(max_recent),
		NULL
	};
	struct istream *input = indexer_send_cmd_with_args(cmd, args);

	alarm(5);
	if ((line = i_stream_read_next_line(input)) == NULL)
		i_fatal("read(indexer) failed: %s", i_stream_get_error(input));
	if (strcmp(line, "0\tOK") != 0)
		i_fatal("indexer: %s returned unexpected reply: %s", cmd, line);
	alarm(0);
	i_stream_destroy(&input);
}

static void cmd_indexer_remove(struct doveadm_cmd_context *cctx)
{
	const char *line, *user_mask, *mailbox_mask;

	if (!doveadm_cmd_param_str(cctx, "user-mask", &user_mask))
		help_ver2(&doveadm_cmd_indexer_remove);
	if (!doveadm_cmd_param_str(cctx, "mailbox-mask", &mailbox_mask))
		mailbox_mask = NULL;

	const char *const args[] = { "0", user_mask, mailbox_mask, NULL };
	struct istream *input = indexer_send_cmd_with_args("REMOVE", args);

	alarm(5);
	if ((line = i_stream_read_next_line(input)) == NULL)
		i_fatal("read(indexer) failed: %s", i_stream_get_error(input));
	if (strcmp(line, "0\tOK") != 0)
		i_fatal("indexer: REMOVE returned unexpected reply: %s", line);
	alarm(0);
	i_stream_destroy(&input);
}

static int cmd_indexer_list_print(const char *const *args)
{
	if (str_array_length(args) < 7)
		return -1;

	/* <tag> <username> <mailbox> <session-id> <max-recent-msgs> <type>
	   <flags> */
	doveadm_print(args[1]);
	doveadm_print(args[2]);
	doveadm_print(args[3]);
	doveadm_print(args[4]);
	switch (args[5][0]) {
	case 'i':
		doveadm_print("index");
		break;
	case 'o':
		doveadm_print("optimize");
		break;
	default:
		doveadm_print(args[5]);
		break;
	}
	if (args[6][0] != 'w')
		doveadm_print("queued");
	else if (args[6][1] == 'h')
		doveadm_print("working/head-queued");
	else if (args[6][1] == 't')
		doveadm_print("working/tail-queued");
	else
		doveadm_print("working");
	return 0;
}

static void cmd_indexer_list(struct doveadm_cmd_context *cctx)
{
	const char *line, *user_mask;

	if (!doveadm_cmd_param_str(cctx, "user-mask", &user_mask))
		user_mask = NULL;

	const char *const args[] = { "0", "all", user_mask, NULL };
	struct istream *input = indexer_send_cmd_with_args("LIST", args);

	doveadm_print_init(DOVEADM_PRINT_TYPE_TABLE);
	doveadm_print_header_simple("username");
	doveadm_print_header_simple("mailbox");
	doveadm_print_header_simple("session_id");
	doveadm_print_header_simple("max_recent");
	doveadm_print_header_simple("type");
	doveadm_print_header_simple("status");

	alarm(30);
	while ((line = i_stream_read_next_line(input)) != NULL) {
		if (strcmp(line, "0") == 0)
			break;
		T_BEGIN {
			const char *const *args = t_strsplit_tabescaped(line);
			if (cmd_indexer_list_print(args) < 0)
				i_fatal("Unexpected input: %s", line);
		} T_END;
	}
	if (line == NULL)
		i_fatal("read(indexer) failed: %s", i_stream_get_error(input));
	alarm(0);
	i_stream_destroy(&input);
}

struct doveadm_cmd_ver2 doveadm_cmd_indexer_add = {
	.cmd = cmd_indexer_add,
	.name = "indexer add",
	.usage = "[-h] [-n <max recent>] <user> <mailbox>",
DOVEADM_CMD_PARAMS_START
DOVEADM_CMD_PARAM('h', "head", CMD_PARAM_BOOL, 0)
DOVEADM_CMD_PARAM('n', "max-recent", CMD_PARAM_STR, 0)
DOVEADM_CMD_PARAM('\0', "user", CMD_PARAM_STR, CMD_PARAM_FLAG_POSITIONAL)
DOVEADM_CMD_PARAM('\0', "mailbox", CMD_PARAM_STR, CMD_PARAM_FLAG_POSITIONAL)
DOVEADM_CMD_PARAMS_END
};

struct doveadm_cmd_ver2 doveadm_cmd_indexer_remove = {
	.cmd = cmd_indexer_remove,
	.name = "indexer remove",
	.usage = "<user mask> [<mailbox mask>]",
DOVEADM_CMD_PARAMS_START
DOVEADM_CMD_PARAM('\0', "user-mask", CMD_PARAM_STR, CMD_PARAM_FLAG_POSITIONAL)
DOVEADM_CMD_PARAM('\0', "mailbox-mask", CMD_PARAM_STR, CMD_PARAM_FLAG_POSITIONAL)
DOVEADM_CMD_PARAMS_END
};

struct doveadm_cmd_ver2 doveadm_cmd_indexer_list = {
	.cmd = cmd_indexer_list,
	.name = "indexer list",
	.usage = "[<user mask>]",
DOVEADM_CMD_PARAMS_START
DOVEADM_CMD_PARAM('\0', "user-mask", CMD_PARAM_STR, CMD_PARAM_FLAG_POSITIONAL)
DOVEADM_CMD_PARAMS_END
};
